plugin/output/loki/loki.go

```diff
@@ -0,0 +1,360 @@
+
+	// > @3@4@5@6
+	// >
+	// > Multiplier for exponential increase of retention between retries
+	RetentionExponentMultiplier int `json:"retention_exponentially_multiplier" default:"2"` // *
+}
+
+type Plugin struct {
+	controller   pipeline.OutputPluginController
+	logger       *zap.SugaredLogger
+	config       *Config
+	avgEventSize int
+
+	httpClient *http.Client
+	batcher    *pipeline.RetriableBatcher
+
+	// plugin metrics
+	sendErrorMetric prometheus.Counter
+}
+
+func init() {
+	fd.DefaultPluginRegistry.RegisterOutput(&pipeline.PluginStaticInfo{
+		Type:    outPluginType,
+		Factory: Factory,
+	})
+}
+
+func Factory() (pipeline.AnyPlugin, pipeline.AnyConfig) {
+	return &Plugin{}, &Config{}
+}
+
+func (p *Plugin) Start(config pipeline.AnyConfig, params *pipeline.OutputPluginParams) {
+	p.controller = params.Controller
+	p.config = config.(*Config)
+	p.logger = params.Logger
+	p.avgEventSize = params.PipelineSettings.AvgEventSize
+
+	batcherOpts := &pipeline.BatcherOptions{
+		PipelineName:   params.PipelineName,
+		OutputType:     outPluginType,
+		Controller:     p.controller,
+		Workers:        p.config.WorkersCount_,
+		BatchSizeCount: p.config.BatchSize_,
+		BatchSizeBytes: p.config.BatchSizeBytes_,
+		FlushTimeout:   p.config.BatchFlushTimeout_,
+		MetricCtl:      params.MetricCtl,
+	}
+
+	backoffOpts := pipeline.BackoffOpts{
+		MinRetention: p.config.Retention_,
+		Multiplier:   float64(p.config.RetentionExponentMultiplier),
+		AttemptNum:   p.config.Retry,
+	}
+
+	onError := func(err error) {
+		var level zapcore.Level
+		if p.config.FatalOnFailedInsert {
+			level = zapcore.FatalLevel
+		} else {
+			level = zapcore.ErrorLevel
+		}
+
+		p.logger.Desugar().Log(level, "can't send data to loki", zap.Error(err),
+			zap.Int("retries", p.config.Retry))
+	}
+
+	p.httpClient = p.newClient(p.config.RequestTimeout_)
+	p.batcher = pipeline.NewRetriableBatcher(
+		batcherOpts,
+		p.out,
+		backoffOpts,
+		onError,
+	)
+
+	p.batcher.Start(context.Background())
+}
+
+func (p *Plugin) Stop() {
+	p.httpClient.CloseIdleConnections()
+	p.batcher.Stop()
+}
+
+func (p *Plugin) Out(event *pipeline.Event) {
+	p.batcher.Add(event)
+}
+
+func (p *Plugin) out(workerData *pipeline.WorkerData, batch *pipeline.Batch) error {
+	if *workerData == nil {
+		*workerData = &data{
+			outBuf: make([]byte, 0, p.config.BatchSize_*p.avgEventSize),
+		}
+	}
+
+	data := (*workerData).(*data)
+
+	// handle too much memory consumption
+	if cap(data.outBuf) > p.config.BatchSize_*p.avgEventSize {
+		data.outBuf = make([]byte, 0, p.config.BatchSize_*p.avgEventSize)
+	}
+
+	root := insaneJSON.Spawn()
+	outBuf := data.outBuf[:0]
+
+	batch.ForEach(func(event *pipeline.Event) {
+		outBuf = root.MutateToNode(event.Root.Node).Encode(outBuf)
+		_ = root.DecodeString("{}")
+	})
+	insaneJSON.Release(root)
+	data.outBuf = outBuf
+
+	p.logger.Debugf("trying to send: %s", outBuf)
+
+	code, err := p.send(context.Background(), outBuf)
+	if err != nil {
+		p.sendErrorMetric.Inc()
```

Comment:
try run with stopped loki

Reply:
![image](https://github.com/user-attachments/assets/24fdcee8-ce12-486a-8ea6-961c3ed225e3)
writing error how it is supposed to do